packages/data-broker/src/services/dhis/DhisInputSchemeResolvingApiProxy.js

```diff
@@ -0,0 +1,244 @@
+
+    const dataElements = await this.models.dataSource.find({ code: query.dataElementCodes });
+
+    for (const dataElement of dataElements) {
+      if (dataElement.config.dhisId) {
+        dataElementIdToCode[dataElement.config.dhisId] = dataElement.code;
+      }
+    }
+
+    return translateElementKeysInEventAnalytics(response, dataElementIdToCode);
+  }
+
+  /**
+   * @param query {*}
+   * @returns bool
+   * @private
+   */
+  allDataElementsHaveDhisId = async query => {
+    const { dataElementCodes } = query;
+
+    const dataElements = await this.models.dataSource.find({ code: dataElementCodes });
+
+    for (const dataElementCode of dataElementCodes) {
+      const dataElement = dataElements.find(d => d.code === dataElementCode);
+
+      if (!dataElement) {
+        // most likely using a data element code that is not defined in data_source, which is supported
+        // by this service
+        return false;
+      }
+
+      if (!dataElement.config.dhisId) {
+        return false;
+      }
+    }
+
+    return true;
+  };
+
+  /**
+   * @param query {*}
+   * @returns bool
+   * @private
+   */
+  allOrgUnitsHaveDhisId = async query => {
+    const { organisationUnitCode, organisationUnitCodes } = query;
+
+    const orgUnitCodes = organisationUnitCode ? [organisationUnitCode] : organisationUnitCodes;
+
+    const mappings = await this.models.dhisOrgUnitMapping.find({ code: orgUnitCodes });
+
+    for (const orgUnitCode of orgUnitCodes) {
+      const mapping = mappings.find(m => m.code === orgUnitCode);
+
+      if (!mapping) {
+        return false;
+      }
+    }
+
+    return true;
+  };
+
+  /**
+   * @param query {*}
+   * @returns bool
+   * @private
+   */
+  allProgramsHaveDhisId = async query => {
+    const { programCode } = query;
+
+    const dataGroups = await this.models.dataSource.find({ code: programCode });
+
+    for (const dataGroup of dataGroups) {
+      if (!dataGroup.config.dhisId) {
+        return false;
+      }
+    }
+
+    return true;
+  };
+
+  /**
+   * @param query {*}
+   * @returns {*}
+   * @private
+   */
+  replaceDataElementCodesWithIds = async query => {
+    const modifiedQuery = { ...query, dataElementIds: [] };
+
+    const { dataElementCodes } = query;
+
+    const dataElements = await this.models.dataSource.find({ code: dataElementCodes });
+
+    for (const dataElementCode of dataElementCodes) {
+      const dataElement = dataElements.find(d => d.code === dataElementCode);
+
+      if (!dataElement) {
+        throw new Error(
+          'DataElement not found in data_source, attempted to replace its code with an id',
+        );
+      }
+
+      if (!dataElement.config.dhisId) {
+        throw new Error(
+          'DataElement does not have a dhisId, attempted to replace its code with the id',
+        );
+      }
+
+      modifiedQuery.dataElementIds.push(dataElement.config.dhisId);
+    }
+
+    delete modifiedQuery.dataElementCodes;
+    return modifiedQuery;
```

Comment:
A more aggressive destructure up front avoids a `delete` later. Up to you though, it's a style thing and is fine either way

```suggestion
    const { dataElementCodes, ...restOfQuery } = query;
    
    const modifiedQuery = { ...restOfQuery, dataElementIds: [] };

    const dataElements = await this.models.dataSource.find({ code: dataElementCodes });

    for (const dataElementCode of dataElementCodes) {
      const dataElement = dataElements.find(d => d.code === dataElementCode);

      if (!dataElement) {
        throw new Error(
          'DataElement not found in data_source, attempted to replace its code with an id',
        );
      }

      if (!dataElement.config.dhisId) {
        throw new Error(
          'DataElement does not have a dhisId, attempted to replace its code with the id',
        );
      }

      modifiedQuery.dataElementIds.push(dataElement.config.dhisId);
    }

    return modifiedQuery;
```

Reply:
Yeah good point, I might leave it in this case but will definitely destructure more aggressively next time